### PR TITLE
add vertex filter for GetNeighbors

### PR DESCRIPTION
**Pull request by nevermore3**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

#### Issue(s) number: 
close https://github.com/vesoft-inc/nebula/issues/4649

#### Description:

Add vertexFilter to getNeighbor interface, when the filter condition of the  vertex is not satisfied, do not return the data for this vertex and corresponding edge 
 

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

**Comment by codesigner** [comment]
Please add some description of what this pr for, from code it seems pushdown vertex filter to storage, while graph side logic seems not changed, only pass nullptr to the new added vertex_filte.

**Comment by nevermore3** [comment]
> Please add some description of what this pr for, from code it seems pushdown vertex filter to storage, while graph side logic seems not changed, only pass nullptr to the new added vertex_filte.

Yes, the getNeightbor interface of the graph side behaves the same as before， it is only used for subgraphs

File: src/storage/query/QueryBaseProcessor-inl.h [hunk 1]
```diff
@@ -152,6 +152,13 @@ nebula::cpp2::ErrorCode QueryBaseProcessor<REQ, RESP>::buildFilter(
     if (filter_ == nullptr) {
       return nebula::cpp2::ErrorCode::E_INVALID_FILTER;
     }
+    const auto* vertexFilterStr = getFilter(req, true);
```
Comment (critical27) [inline src/storage/query/QueryBaseProcessor-inl.h:155]: Be careful here, if the filter is vertex-only, both filter_ and vertexFilter_ will be set, the Filter will eval more than once?

**Review by critical27 — COMMENTED** [review]
Is it a temporary solution for https://github.com/vesoft-inc/nebula/issues/4649?

**Comment by nevermore3** [comment]
> Is it a temporary solution for #4649?

After discussing with Peng Wei, this is the final solution

File: src/storage/query/QueryBaseProcessor-inl.h (hunk 1 above)
Comment (nevermore3) [reply src/storage/query/QueryBaseProcessor-inl.h:155]: Yes, it will be executed twice if vertexFilter_ and filter_ are the same And the result is true

Comment (critical27) [reply src/storage/query/QueryBaseProcessor-inl.h:155]: Is `filter` and `vertex-filter` mutually exclusive? (can't set them both at the same request)? If is, we only need to eval once I think~

BTW, where do we extract vertex filter? I didn't see the code to set it.

File: src/storage/exec/FilterNode.h [hunk 2]
```diff
@@ -68,6 +70,10 @@ class FilterNode : public IterateNode<T> {
     mode_ = mode;
   }
 
+  void setVertexFilter(Expression* vertexFilter) {
```
Comment (pengweisong) [inline src/storage/exec/FilterNode.h:73]: why not set it through constructor?

File: src/storage/exec/GetNeighborsNode.h [hunk 3]
```diff
@@ -50,6 +50,10 @@ class GetNeighborsNode : public QueryNode<VertexID> {
       return nebula::cpp2::ErrorCode::E_INVALID_DATA;
     }
 
+    if (context_->resultStat_ == ResultStatus::TAG_FILTER_OUT) {
```
Comment (pengweisong) [inline src/storage/exec/GetNeighborsNode.h:53]: could add some comments.

File: src/storage/exec/FilterNode.h [hunk 4]
```diff
@@ -93,6 +99,13 @@ class FilterNode : public IterateNode<T> {
   // return true when the value iter points to a value which can filter
   bool checkTagAndEdge() {
     expCtx_->reset(this->reader(), this->key().str());
+    if (vertexFilterExp_ != nullptr) {
```
Comment (pengweisong) [inline src/storage/exec/FilterNode.h:102]: should this be also evaluated in `checkTagOnly`?

File: src/storage/query/GetNeighborsProcessor.cpp [hunk 5]
```diff
@@ -313,13 +316,16 @@ nebula::cpp2::ErrorCode GetNeighborsProcessor::checkAndBuildContexts(
...
   }
-  code = buildFilter(req, [](const cpp2::GetNeighborsRequest& r) -> const std::string* {
-    if (r.get_traverse_spec().filter_ref().has_value()) {
-      return r.get_traverse_spec().get_filter();
-    } else {
-      return nullptr;
-    }
-  });
+  code =
+      buildFilter(req, [](const cpp2::GetNeighborsRequest& r, bool isVertex) -> const std::string* {
```
Comment (pengweisong) [inline src/storage/query/GetNeighborsProcessor.cpp:320]: `onlyVertex` may be better then `isVertex`

File: src/storage/query/QueryBaseProcessor-inl.h [hunk 6]
```diff
@@ -152,6 +152,13 @@ nebula::cpp2::ErrorCode QueryBaseProcessor<REQ, RESP>::buildFilter(
...
+    if (vertexFilterStr != nullptr && !vertexFilterStr->empty()) {
+      vertexFilter_ = Expression::decode(pool, *vertexFilterStr);
```
Comment (pengweisong) [inline src/storage/query/QueryBaseProcessor-inl.h:157]: should `vertexFilter_` be check in `checkExp`?

File: src/storage/exec/FilterNode.h (hunk 2 above)
Comment (nevermore3) [reply src/storage/exec/FilterNode.h:73]: fixed

File: src/storage/query/QueryBaseProcessor-inl.h (hunk 6 above)
Comment (nevermore3) [reply src/storage/query/QueryBaseProcessor-inl.h:157]: vertexfilter is a subset of filter_ and does not need to be checked repeatedly

**Review by critical27 — APPROVED** [review]

**Review by pengweisong — APPROVED** [review]
Good job

**Pull request merged**